### PR TITLE
Add Windows Crash Dump format support

### PR DESCRIPTION
**Pull request by abcSup**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [x] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**

This PR includes a r_bin plugin to analyze Windows Crash Dump (.dmp) files. The current implementation parses the following information:
* 64bit .dmp header
* BMP header
* Physical Memory Runs Pages
* Bitmap Pages from BMP header
* `vtop` Address Translation for AMD64 Paged Memory (implemented with r_list) (unused)

After working on it, I realized that it requires more than what [kdmp-parser](https://github.com/0vercl0k/kdmp-parser) to extract extra information such as those presented by dumpchk.exe or WinDbg:
* Loaded modules/drivers, Kernel Base,`PsloadedModuleList`
* Stack backtrace

Since the kernel dump contains the physical memory blocks, we would need a better Address Translation method to cater for different page tables, such as the one in [rekall](https://github.com/google/rekall/blob/c5d68e31705f4b5bd2581c1d951b7f6983f7089c/rekall-core/rekall/plugins/addrspaces/amd64.py#L47). I am going to work on the debugger side of radare2, so I stopped right here.

**Test plan**

I am going to upload a small kernel dump (85MB) to the testbins and test header parsing and pages mapping. However, a full dump will have at least 1GB which will greatly slows down all the tests in future, so I might skip that (?)

**Closing issues**

#16048 (There are many more improvements needed actually.)

### CODE REVIEW

**Comment by ITAYC0HEN** [comment]
Good work :)

**Comment by XVilka** [comment]
```
[XX] db\formats\dmp\dmp : DMP64 Check  478
$ r2 -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Q -i C:\Users\appveyor\AppData\Local\Temp\1\tmp-2920a4b5FoJkebyc ../bins/dmp/bmp.dmp
i~format
-format   dmp64
+format   any
EXPECT='format   any'
[XX] db\formats\dmp\dmp : Resolve vaddr  490
$ r2 -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Q -i C:\Users\appveyor\AppData\Local\Temp\1\tmp-2920b9ktO1Ka5NvR ../bins/dmp/bmp.dmp
px 64 @ 0x1000
 - offset -   0 1  2 3  4 5  6 7  8 9  A B  C D  E F  0123456789ABCDEF
-0x00001000  0000 0000 0000 0000 0000 f9ff 0000 0000  ................
-0x00001010  0006 0101 0000 0000 0000 0000 0000 0000  ................
-0x00001020  0000 0000 0000 0000 0000 0000 0000 0000  ................
-0x00001030  0000 0000 0000 0000 0000 0000 0000 0000  ................
+0x00001000  5041 4745 5041 4745 5041 4745 5041 4745  PAGEPAGEPAGEPAGE
+0x00001010  5041 4745 5041 4745 5041 4745 5041 4745  PAGEPAGEPAGEPAGE
+0x00001020  5041 4745 5041 4745 5041 4745 5041 4745  PAGEPAGEPAGEPAGE
+0x00001030  9d0a 9d2a 0000 0000 5041 4745 5041 4745  ...*....PAGEPAGE
EXPECT='- offset -   0 1  2 3  4 5  6 7  8 9  A B  C D  E F  0123456789ABCDEF
0x00001000  5041 4745 5041 4745 5041 4745 5041 4745  PAGEPAGEPAGEPAGE
0x00001010  5041 4745 5041 4745 5041 4745 5041 4745  PAGEPAGEPAGEPAGE
0x00001020  5041 4745 5041 4745 5041 4745 5041 4745  PAGEPAGEPAGEPAGE
0x00001030  9d0a 9d2a 0000 0000 5041 4745 5041 4745  ...*....PAGEPAGE'
[XX] db\formats\dmp\dmp : DMP BMP Page sections  649
$ r2 -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Q -i C:\Users\appveyor\AppData\Local\Temp\1\tmp-2920NzN96VcQ3R9G ../bins/dmp/bmp.dmp
iS~?
-21582
+3
EXPECT='3'
[XX] db\formats\dmp\dmp : DMP64 Info  742
$ r2 -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Q -i C:\Users\appveyor\AppData\Local\Temp\1\tmp-2920E0gFVEmQCZxn ../bins/dmp/bmp.dmp
iI
-arch     x86
-baddr    0xffffffffffffffff
+baddr    0x0
 binsz    88428600
-bintype  dmp64
 bits     64
 canary   false
-retguard false
 crypto   false
 endian   little
-havecode true
+havecode false
 laddr    0x0
 linenum  false
 lsyms    false
-machine  AMD64
 maxopsz  16
 minopsz  1
 nx       false
-os       Unknown
 pcalign  0
 pic      false
 relocs   false
 sanitiz  false
 static   true
 stripped false
-va       true
+va       false
EXPECT='baddr    0x0
binsz    88428600
bits     64
canary   false
crypto   false
endian   little
havecode false
laddr    0x0
linenum  false
lsyms    false
maxopsz  16
minopsz  1
nx       false
pcalign  0
pic      false
relocs   false
sanitiz  false
static   true
stripped false
va       false'
[XX] db\formats\dmp\dmp : DMP header & bmp header  747
$ r2 -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Q -i C:\Users\appveyor\AppData\Local\Temp\1\tmp-2920nk1zbk57vnhD ../bins/dmp/bmp.dmp
iH
-DUMP_HEADER64:
-  MajorVersion : 0x0000000f
-  MinorVersion : 0x000047ba
-  DirectoryTableBase : 0x00000000006d4002
-  PfnDataBase : 0xfffff80510c22468
-  PsLoadedModuleList : 0xfffff80510af62f0
-  PsActiveProcessHead : 0xfffff80510ae6980
-  MachineImageType : 0x00008664
-  NumberProcessors : 0x00000001
-  BugCheckCode : 0x00000000
-  BugCheckParameter1 : 0x0000000000000000
-  BugCheckParameter2 : 0x0000000000000000
-  BugCheckParameter3 : 0x0000000000000000
-  BugCheckParameter4 : 0x0000000000000000
-  KdDebuggerDataBlock : 0xfffff80510ad45e0
-  SecondaryDataState : 0x45474150
-  ProductType : 0x45474150
-  SuiteMask : 0x45474150
-
-BITMAP_DUMP:
-  HeaderSize : 0x0000a038
-  BitmapSize : 0x00040000
-  Pages : 0x0000544b
EXPECT=''
[**]             db\formats\dmp\dmp   OK 1807 BR   52 XX    5 FX    7
```

**Comment by radare** [comment]
lgtm, just add the required lines in the meson.build files, the failing tests are because this plugin is not build with meson

**Comment by radare** [comment]
ping! rls is today1

**Comment by abcSup** [comment]
Thanks everyone! I just got up.

File: libr/bin/format/dmp/dmp64.c [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+
+static int r_bin_dmp64_init_memory_runs(struct r_bin_dmp64_obj_t *obj) {
+	int i, j;
+	dmp64_p_memory_desc *mem_desc = &obj->header->PhysicalMemoryBlockBuffer;
+	if (!memcmp (mem_desc, DMP_UNUSED_MAGIC, 4)) {
+		eprintf ("Warning: Invalid PhysicalMemoryDescriptor\n");
+		return false;
+	}
+	ut64 num_runs = mem_desc->NumberOfRuns;
+	if (num_runs * sizeof (dmp_p_memory_run) >= 0x348) {
```
Comment (ret2libc) [inline libr/bin/format/dmp/dmp64.c]: what is this magic number?

Comment (ret2libc) [reply libr/bin/format/dmp/dmp64.c]: Probably you should DEFINE it somewhere with a meaningful name.

File: libr/bin/format/dmp/dmp64.c [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+		eprintf ("Warning: Invalid PhysicalMemoryDescriptor\n");
+		return false;
+	}
+	obj->pages = r_list_newf (free);
+	if (!obj->pages) {
+		return false;
+	}
+	dmp_p_memory_run *runs = calloc (num_runs, sizeof (dmp_p_memory_run));
+	if (r_buf_read_at (obj->b, 0x98, (ut8*)runs, num_runs * sizeof (dmp_p_memory_run)) < 0) {
```
Comment (ret2libc) [inline libr/bin/format/dmp/dmp64.c]: this other magic number

File: libr/bin/format/dmp/dmp64.c (hunk 1 above)
Comment (abcSup) [reply libr/bin/format/dmp/dmp64.c]: That is the offset to the next property. That line is to check if the PhysicalMemoryBlockBuffer has valid size. Is there a way to easily retrieve a specific offset of a item in a structure without DEFINE?
```
0x80  : KdDebuggerDataBlock            ['unsigned long long']
0x88  : PhysicalMemoryBlockBuffer      ['_PHYSICAL_MEMORY_DESCRIPTOR']
0x348 : ContextRecord                  ['array', 3000, ['unsigned char']]
0xf00 : Exception                      ['_EXCEPTION_RECORD64']
```

File: libr/bin/format/dmp/dmp64.c (hunk 2 above)
Comment (abcSup) [reply libr/bin/format/dmp/dmp64.c]: Offset to NumberOfPages property in relative to the header base (0x0).

```
0x88  : PhysicalMemoryBlockBuffer      ['_PHYSICAL_MEMORY_DESCRIPTOR']
```
```
typedef struct _PHYSICAL_MEMORY_DESCRIPTOR64 {
	ut32 NumberOfRuns; // 0x0
	ut32 _padding1;
	ut64 NumberOfPages; // 0x8
	dmp_p_memory_run Run[1];
} dmp64_p_memory_desc;
```

File: libr/bin/format/dmp/dmp64.c (hunk 1 above)
Comment (abcSup) [reply libr/bin/format/dmp/dmp64.c]: DEFINE would be fine, just wondering if there is another way without doing something like, `&(obj->header->NumberOfPages) - obj->header`

Comment (abcSup) [reply libr/bin/format/dmp/dmp64.c]: Nvm, I found r_offsetof!

**Comment by radare** [comment]
lgtm now!

**Pull request merged**